File: android/src/main/java/expo/modules/cedarling/CedarlingResultMapper.kt
package expo.modules.cedarling

import uniffi.cedarling_uniffi.AuthorizeResult
import uniffi.cedarling_uniffi.Decision
import uniffi.cedarling_uniffi.Diagnostics
import uniffi.cedarling_uniffi.MultiIssuerAuthorizeResult

internal object CedarlingResultMapper {
  fun map(result: AuthorizeResult): Map<String, Any> =
    map(result.decision, result.response.decision, result.requestId, result.response.diagnostics)

  fun map(result: MultiIssuerAuthorizeResult): Map<String, Any> =
    map(result.decision, result.response.decision, result.requestId, result.response.diagnostics)

  internal fun map(
    allowed: Boolean,
    decision: Decision,
    requestId: String,
    diagnostics: Diagnostics
  ): Map<String, Any> {
    val enumAllows = decision == Decision.ALLOW
    if (allowed != enumAllows) {
      throw CedarlingSdkException(
        CedarlingErrorCode.NATIVE_RESULT_INCONSISTENT,
        "native Boolean and Cedar decision disagree"
      )
    }
    if (requestId.isBlank()) {
      throw CedarlingSdkException(
        CedarlingErrorCode.NATIVE_RESULT_INCONSISTENT,
        "native result has an empty request ID"
      )
    }
    return mapOf(
      "allowed" to allowed,
      "decision" to decision.name,
      "requestId" to requestId,
      "diagnostics" to mapOf(
        "reasons" to diagnostics.reasons,
        "errors" to diagnostics.errors
      )
    )
  }
}

File: ios/CedarlingResultMapper.swift
import Foundation

internal enum CedarlingResultMapper {
  static func map(_ result: AuthorizeResult) throws -> [String: Any] {
    try map(
      allowed: result.decision,
      decision: result.response.decision,
      requestId: result.requestId,
      diagnostics: result.response.diagnostics
    )
  }

  static func map(_ result: MultiIssuerAuthorizeResult) throws -> [String: Any] {
    try map(
      allowed: result.decision,
      decision: result.response.decision,
      requestId: result.requestId,
      diagnostics: result.response.diagnostics
    )
  }

  static func map(
    allowed: Bool,
    decision: Decision,
    requestId: String,
    diagnostics: Diagnostics
  ) throws -> [String: Any] {
    let decisionName: String
    switch decision {
    case .allow:
      decisionName = "ALLOW"
    case .deny:
      decisionName = "DENY"
    }

    guard allowed == (decisionName == "ALLOW") else {
      throw CedarlingModuleError(
        .nativeResultInconsistent,
        "native Boolean and Cedar decision disagree"
      )
    }
    guard !requestId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      throw CedarlingModuleError(
        .nativeResultInconsistent,
        "native result has an empty request ID"
      )
    }
    return [
      "allowed": allowed,
      "decision": decisionName,
      "requestId": requestId,
      "diagnostics": [
        "reasons": Array(diagnostics.reasons),
        "errors": Array(diagnostics.errors)
      ]
    ]
  }
}
